src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/SourceGeneratorProjectEngine.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using Microsoft.AspNetCore.Razor.Language;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.NET.Sdk.Razor.SourceGenerators;
+
+internal class SourceGeneratorProjectEngine : DefaultRazorProjectEngine
+{
+	private readonly int discoveryPhaseIndex = -1;
+
+	private readonly int rewritePhaseIndex = -1;
+
+	public SourceGeneratorProjectEngine(DefaultRazorProjectEngine projectEngine)
+		: base(projectEngine.Configuration, projectEngine.Engine, projectEngine.FileSystem, projectEngine.ProjectFeatures)
+    {
+		for(int i = 0; i < Engine.Phases.Count; i++)
+		{
+			if (Engine.Phases[i] is DefaultRazorTagHelperContextDiscoveryPhase)
+			{
+				discoveryPhaseIndex = i;
+			}
+			else if (Engine.Phases[i] is DefaultRazorTagHelperRewritePhase)
+			{
+				rewritePhaseIndex = i;
+			}
+			else if (discoveryPhaseIndex >= 0 && rewritePhaseIndex >= 0)
+			{
+				break;
+			}
+		}
+		Debug.Assert(discoveryPhaseIndex >= 0);
+		Debug.Assert(rewritePhaseIndex >= 0);
+	}
+
+	public SourceGeneratorRazorCodeDocument ProcessInitialParse(RazorProjectItem projectItem)
+	{
+		var codeDocument = CreateCodeDocumentCore(projectItem);
+		ProcessPartial(codeDocument, 0, discoveryPhaseIndex);
+
+		// record the syntax tree, before the tag helper re-writing occurs
+		codeDocument.SetPreTagHelperSyntaxTree(codeDocument.GetSyntaxTree());
+		return new SourceGeneratorRazorCodeDocument(codeDocument);
+	}
+
+	public SourceGeneratorRazorCodeDocument ProcessTagHelpers(SourceGeneratorRazorCodeDocument sgDocument, IReadOnlyList<TagHelperDescriptor> tagHelpers, bool checkForIdempotency)
+	{
+		Debug.Assert(sgDocument.CodeDocument.GetPreTagHelperSyntaxTree() is not null);
+
+		int startIndex = discoveryPhaseIndex;
+		var codeDocument = sgDocument.CodeDocument;
+		var previousTagHelpers = codeDocument.GetTagHelpers();
+		if (checkForIdempotency && previousTagHelpers is not null)
```

Comment:
This is where we check the idempotency of the tag helpers. If they haven't changed, we can skip re-writing. If they have, we can check if the ones that changed can actually have an impact on the document and skip if not.